core/arch/arm/kernel/generic_boot.c

```diff
@@ -471,17 +471,67 @@ static void init_runtime(unsigned long pageable_part __unused)
 }
 #endif
 
-#if defined(CFG_DT) && !defined(CFG_EMBED_DTB)
-void *get_dt_blob(void)
+#if defined(CFG_DT)
+#if defined(CFG_EMBED_DTB)
+static void *get_embedded_dt_blob(void)
 {
+	static bool checked;
+
 	assert(cpu_mmu_enabled());
-	return dt_blob_addr;
+
+	if (!checked) {
+		if (fdt_check_header(embedded_secure_dtb))
+			panic("Invalid embedded DTB");
+
+		checked = true;
+	}
+
+	return embedded_secure_dtb;
+}
+#else
+static void *get_embedded_dt_blob(void)
+{
+	return NULL;
+}
+#endif /*CFG_EMBED_DTB*/
+
+void *get_sec_dtb(void)
+{
+	void *fdt = get_embedded_dt_blob();
+
+	if (fdt)
+		return fdt;
+
+	if (tee_vbuf_is_sec(external_dt_blob, CFG_DTB_MAX_SIZE))
+		return external_dt_blob;
+
+	return NULL;
+}
+
+void *get_ns_dtb(void)
+{
+	assert(cpu_mmu_enabled());
+
+	if (tee_vbuf_is_non_sec(external_dt_blob, CFG_DTB_MAX_SIZE))
```

Comment:
I think this is what is breaking QEMU (Travis CI failure). It seems that plat-vexpress maps everything secure (try "git grep NSEC" in `core/arch/arm/plat-vexpress` => no match). Therefore the optee node is not added to the NW DT.

I think our best option is to remove this check and relax a bit the specification of this function: "Returns the device tree used by Normal World".

Thoughts?

Reply:
The mapping is done from **generic_boot.c**, not from the platform ([init_fdt()](https://github.com/OP-TEE/optee_os/pull/2673/files/32a525a7a18341c51142cc09de281dfb9cbfbf9c#diff-91935058a1bd4fa07820044637b0d980R848)). I should not fail here.

Note: "Returns the device tree used by Normal World" implicitly expect DTB is located in non-secure memory.

Reply:
> Note: "Returns the device tree used by Normal World" implicitly expect DTB is located in non-secure memory.

The bootloader could move it, we don't know.

Reply:
I'm debugging the QEMU problem with this PR now, I'll get back with the result.

Reply:
There seems to be a chicken-egg problem. QEMU doesn't define nsec DDR instead it discovers it from the DTB.
The following patch takes care of that, but then we end up with secure and normal world sharing console.

```patch
--- a/core/arch/arm/kernel/generic_boot.c
+++ b/core/arch/arm/kernel/generic_boot.c
@@ -910,7 +910,10 @@ static void discover_nsec_memory(void)
 {
 	struct core_mmu_phys_mem *mem;
 	size_t nelems;
-	void *fdt = get_dtb();
+	void *fdt = get_embedded_dt_blob();
+
+	if (!fdt)
+		fdt = external_dt_blob;
 
 	if (fdt) {
 		mem = get_memory(fdt, &nelems);
@@ -957,9 +960,9 @@ static void init_primary_helper(unsigned long pageable_part,
 	thread_init_per_cpu();
 	init_sec_mon(nsec_entry);
 	init_fdt(fdt);
+	discover_nsec_memory();
 	update_fdt();
 	configure_console_from_dt();
-	discover_nsec_memory();
 
 	IMSG("OP-TEE version: %s", core_v_str);

Reply:
ok, thanks.

Reply:
@jenswi-linaro 

> QEMU doesn't define nsec DDR, instead it discovers it from the DTB.

You mean OP-TEE (for QEMU) doesn't define... right?

> The following patch takes care of that, but then we end up with secure and normal world sharing console.

Unless I'm missing something, that should addressed by switching to a more recent version of QEMU, see https://github.com/OP-TEE/manifest/pull/125.

Reply:
Yes, I mean OP-TEE for QEMU.

OK, I'll upgrade QEMU

Reply:
@jforissier : successfully tested your manifest with and without this P-R. thanks. I think we should wait **default.xml** is updated with your changes before we merge this P-R. In the meantime, i will rebase this P-R on top of https://github.com/OP-TEE/optee_os/pull/2666 that seems ready-to-be-merged.

Reply:
@etienne-lms sounds like a plan! Thanks.